bindings/sundials4py/idas/idas.cpp

```diff
@@ -0,0 +1,319 @@
+        {
+          auto fn_table    = get_idas_fn_table(ida_mem);
+          fn_table->rootfn = nb::cast(fn);
+          return IDARootInit(ida_mem, nrtfn, &idas_rootfn_wrapper);
+        });
+
+  m.def("IDAQuadInit",
+        [](void* ida_mem,
+           std::function<std::remove_pointer_t<IDAQuadRhsFn>> resQ, N_Vector yQ0)
+        {
+          auto fn_table  = get_idas_fn_table(ida_mem);
+          fn_table->resQ = nb::cast(resQ);
+          return IDAQuadInit(ida_mem, &idas_resQ_wrapper, yQ0);
+        });
+
+  BIND_IDA_CALLBACK(IDAWFtolerances, IDAEwtFn, ewtn, idas_ewtfn_wrapper,
+                    nb::arg("ida_mem"), nb::arg("efun").none());
+
+  BIND_IDA_CALLBACK(IDASetNlsResFn, IDAResFn, resNLS, idas_nlsresfn_wrapper,
+                    nb::arg("ida_mem"), nb::arg("res").none());
+
+  BIND_IDA_CALLBACK(IDASetJacFn, IDALsJacFn, lsjacfn, idas_lsjacfn_wrapper,
+                    nb::arg("ida_mem"), nb::arg("jac").none());
+
+  BIND_IDA_CALLBACK2(IDASetPreconditioner, IDALsPrecSetupFn, lsprecsetupfn,
+                     idas_lsprecsetupfn_wrapper, IDALsPrecSolveFn, lsprecsolvefn,
+                     idas_lsprecsolvefn_wrapper, nb::arg("ida_mem"),
+                     nb::arg("pset").none(), nb::arg("psolve").none());
+
+  BIND_IDA_CALLBACK2(IDASetJacTimes, IDALsJacTimesSetupFn, lsjactimessetupfn,
+                     idas_lsjactimessetupfn_wrapper, IDALsJacTimesVecFn,
+                     lsjactimesvecfn, idas_lsjactimesvecfn_wrapper,
+                     nb::arg("ida_mem"), nb::arg("jtsetup").none(),
+                     nb::arg("jtimes").none());
+
+  BIND_IDA_CALLBACK(IDASetJacTimesResFn, IDALsJacTimesVecFn, lsjacresfn,
+                    idas_lsjacresfn_wrapper, nb::arg("ida_mem"),
+                    nb::arg("jtimesResFn").none());
+
+  //
+  // Sensitivity and quadrature sensitivity bindings
+  //
+
+  m.def("IDAQuadSensInit",
+        [](void* ida_mem, std::function<IDAQuadSensRhsStdFn> resQS,
+           std::vector<N_Vector> yQS0)
+        {
+          auto fn_table   = get_idas_fn_table(ida_mem);
+          fn_table->resQS = nb::cast(resQS);
+          return IDAQuadSensInit(ida_mem, idas_resQS_wrapper, yQS0.data());
+        });
+
+  m.def("IDASensInit",
+        [](void* ida_mem, int Ns, int ism, std::function<IDASensResStdFn> resS,
+           std::vector<N_Vector> yS0, std::vector<N_Vector> ypS0)
+        {
+          auto fn_table  = get_idas_fn_table(ida_mem);
+          fn_table->resS = nb::cast(resS);
+          return IDASensInit(ida_mem, Ns, ism, idas_resS_wrapper, yS0.data(),
+                             ypS0.data());
+        });
+
+  ///
+  // IDAS adjoint bindings
+  ///
+
+  m.def("IDAInitB",
+        [](void* ida_mem, int which,
+           std::function<std::remove_pointer_t<IDAResFnB>> resB,
+           sunrealtype tB0, N_Vector yyB0, N_Vector ypB0)
+        {
+          int ida_status = IDAInitB(ida_mem, which, idas_resB_wrapper, tB0,
+                                    yyB0, ypB0);
```

Comment:
Since we always pass `ida_resB_wrapper` to `IDAInit` check if the residual is `NULL` and check return flag
```suggestion
          if (!resB) { throw sundials4py::illegal_value("resB was null"); }
          int ida_status = IDAInitB(ida_mem, which, idas_resB_wrapper, tB0,
                                    yyB0, ypB0);
          if (ida_status != IDA_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize IDA");
          }
```
Should we return the error flag?